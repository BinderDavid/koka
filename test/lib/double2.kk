/*

*/

public import std/num/double


fun check(name : string, res : string, got : string ) : io () {
  println(name.pad-right(7,' ') + ": " 
    + (if (got == res) then "ok: " + res 
                       else "FAILED!:\n expect: " + res + "\n gotten: " + got + "\n"))
  // if (got != res) then error("failed") else ()
}

fun check2( name : string, res : string, tst : (double, int) ) : io () {
  check(name,res, tst.fst.show + ", " + tst.snd.show)
}

fun check1(name : string, res : string, tst : double ) : io () {
  check(name,res,tst.show)
}

fun check-logp(name : string, res : string, x : double ) : io () {
  check(name,res, log(x + 1.0).show + ", " + log1p(x).show + ", " + jslog1p(x).show)
}

extern jslog1p( x : double ) : double {
  js "Math.log1p"
}

fun main() {
  check-logp("logp 1", "0.6931471805599453, 0.6931471805599453, 0.6931471805599453", 1.0)
  check-logp("logp 2", "1.332267629550187e-15, 1.2344999999999995e-15, 1.2344999999999993e-15", 1.2345e-15)
  check-logp("logp 3", "2.2204460492503128e-16, 1.2344999999999996e-16, 1.2345e-16", 1.2345e-16)
  check-logp("logp 4", "0, 1.2345e-32, 1.2345e-32", 1.2345e-32)
  check-logp("logp 5", "0, 1.2345e-64, 1.2345e-64", 1.2345e-64)
  check-logp("logp 6", "709.782712893384, 709.782712893384, 709.782712893384", 1.7976931348622732e+308)
}