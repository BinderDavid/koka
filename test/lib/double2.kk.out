logp 1 : ok: 0.6931471805599453, 0.6931471805599453, 0.6931471805599453
logp 2 : ok: 1.332267629550187e-15, 1.2344999999999995e-15, 1.2344999999999993e-15
logp 3 : ok: 2.2204460492503128e-16, 1.2344999999999996e-16, 1.2345e-16
logp 4 : ok: 0, 1.2345e-32, 1.2345e-32
logp 5 : ok: 0, 1.2345e-64, 1.2345e-64
logp 6 : ok: 709.782712893384, 709.782712893384, 709.782712893384
atanh: 0: ok: 0.0000000000000
atanh: 1.23e-16: ok: 2.2204460492503e-16
atanh: 1.23e-15: ok: 1.3322676295502e-15
atanh: 0.00123: ok: 0.0012300006202896
atanh: 0.123: ok: 0.12362598118313
atanh: 0.5: ok: 0.54930614433405
atanh: 0.6931471805599453: ok: 0.85398804799752
atanh: 1: ok: Infinity
atanh: 2: ok: NaN
atanh: Infinity: ok: NaN
atanh: NaN: ok: NaN
atanh: 0: ok: 0.0000000000000
atanh: -1.23e-16: ok: -1.6653345369377e-16
atanh: -1.23e-15: ok: -1.2767564783189e-15
atanh: -0.00123: ok: -0.0012300006202896
atanh: -0.123: ok: -0.12362598118313
atanh: -0.5: ok: -0.54930614433405
atanh: -0.6931471805599453: ok: -0.85398804799752
atanh: -1: ok: -Infinity
atanh: -2: ok: NaN
atanh: -Infinity: ok: NaN
atanh: NaN: ok: NaN
asinh: 0: ok: 0.0000000000000
asinh: 1.23e-16: FAILED!:
 expect: 1.2300000000000e-16
 gotten: 2.2204460492503e-16
asinh: 1.23e-15: FAILED!:
 expect: 1.2300000000000e-15
 gotten: 1.3322676295502e-15
asinh: 0.00123: FAILED!:
 expect: 0.0012299996898557
 gotten: 0.0012299996898556
asinh: 0.122: ok: 0.12169936791920
asinh: 0.5: ok: 0.48121182505960
asinh: 0.6931471805599453: ok: 0.64704348108319
asinh: 1: ok: 0.88137358701954
asinh: 2: ok: 1.4436354751788
asinh: 1230: ok: 7.8079167941719
asinh: 123000000: ok: 19.320842093897
asinh: 1230000000000000: ok: 35.438937744855
asinh: 12300000000000000: ok: 37.741522837849
asinh: Infinity: ok: Infinity
asinh: NaN: ok: NaN
asinh: 0: ok: 0.0000000000000
asinh: -1.23e-16: FAILED!:
 expect: -1.2300000000000e-16
 gotten: -2.2204460492503e-16
asinh: -1.23e-15: FAILED!:
 expect: -1.2300000000000e-15
 gotten: -1.3322676295502e-15
asinh: -0.00123: FAILED!:
 expect: -0.0012299996898557
 gotten: -0.0012299996898556
asinh: -0.122: ok: -0.12169936791920
asinh: -0.5: ok: -0.48121182505960
asinh: -0.6931471805599453: ok: -0.64704348108319
asinh: -1: ok: -0.88137358701954
asinh: -2: ok: -1.4436354751788
asinh: -1230: ok: -7.8079167941719
asinh: -123000000: ok: -19.320842093897
asinh: -1230000000000000: ok: -35.438937744855
asinh: -12300000000000000: ok: -37.741522837849
asinh: -Infinity: ok: -Infinity
asinh: NaN: ok: NaN
acosh: 0: ok: NaN
acosh: 1.23e-16: ok: NaN
acosh: 1.23e-15: ok: NaN
acosh: 0.00123: ok: NaN
acosh: 0.122: ok: NaN
acosh: 0.5: ok: NaN
acosh: 0.6931471805599453: ok: NaN
acosh: 1: ok: 0.0000000000000
acosh: 2: ok: 1.3169578969248
acosh: 1230: ok: 7.8079164636808
acosh: 123000000: ok: 19.320842093897
acosh: 1230000000000000: ok: 35.438937744855
acosh: 12300000000000000: ok: 37.741522837849
acosh: Infinity: ok: Infinity
acosh: NaN: ok: NaN
acosh: 0: ok: NaN
acosh: -1.23e-16: ok: NaN
acosh: -1.23e-15: ok: NaN
acosh: -0.00123: ok: NaN
acosh: -0.122: ok: NaN
acosh: -0.5: ok: NaN
acosh: -0.6931471805599453: ok: NaN
acosh: -1: ok: NaN
acosh: -2: ok: NaN
acosh: -1230: ok: NaN
acosh: -123000000: ok: NaN
acosh: -1230000000000000: ok: NaN
acosh: -12300000000000000: ok: NaN
acosh: -Infinity: ok: NaN
acosh: NaN: ok: NaN